ENH: stats.f_oneway: add `equal_var` for Welch ANOVA

YongcaiHuang opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
https://github.com/scipy/scipy/issues/11122

#### What does this implement/fix?
<!--Please explain your changes.-->
add `equal_var` argument in `spicy.stats.f_oneway` to perform Welch ANOVA (similar to `spicy.stats.tttest_ind` implementation with `equal_var` argument added)

#### Additional information
<!--Any additional information you think is important.-->
although statsmodels contains welch anova, it is more simple to be used within scipy

refer to 
```
B. L. Welch, “On the Comparison of Several Mean Values: An Alternative Approach,”
         Biometrika, vol. 38, no. 3/4, pp. 330–336, 1951, doi: 10.2307/2332579.

 D. C. Howell, Statistical Methods for Psychology. in International student edition.
         Wadsworth Cengage Learning, 2013.
```

and with benchmark from `oneway.test(var.equal = FALSE)` in R

YongcaiHuang committed:
ENH: f_oneway enhancement with equal_var for welch anova

YongcaiHuang committed:
ENH: f_oneway enhancement with equal_var for welch anova

mdhaber commented:
@josef-pkt Thoughts on this? I closed the linked issue because you had already added this to statsmodels.

YongcaiHuang committed:
Merge branch 'scipy:main' into main

YongcaiHuang committed:
STY: style fix in welch anova docstring

YongcaiHuang committed:
Merge branch 'scipy:main' into main

YongcaiHuang committed:
Merge branch 'scipy:main' into main

YongcaiHuang commented:
Hi this implementation is more about a preparation of `games_howell` PR https://github.com/scipy/scipy/pull/22185

josef-pkt commented:
I will look at it (I was mostly offline in the last 2 weeks)

In general, it's good to have Welsh anova version together with the standard version in scipy to increase the popularity of the version with less restrictive assumptions (being robust to heteroscedasticity)
(similar to t-test)

josef-pkt commented:
looks fine overall, I did not check the details of the computation especially axis handling.
axis handling might need additional unit tests.
(AFAIR, statsmodels anova is not vectorized)

One possible issue:
The docstring mentions pretesting for equal variances, although it does not formulate it as a recommendation.
In general, variance tests in small samples are very noisy and are not recommended as pretest estimators.
I know some of the literature in pre-testing for t-test, but not for the oneway anova.
The frequent recommendation for t-test is to use unequal variance Welch version by default or if there is some prior suspicion that variances are not equal, but not to rely on a rejection decision of a pre-test.

Results are the same as in statsmodels. 
But I managed to use unequal variance/Welch anova as default because we did not have a prior equal variance anova function.
Sstatsmodels has some hypothesis test with non-optimal or not recommended defaults for historical reasons, including tukey-hsd as default, instead of games-howell, or equal var in t-test. I would not choose those defaults now.) 

```
import statsmodels.stats.oneway as ow
from statsmodels.sandbox.stats.multicomp import catstack
samples = [[-50.42, 40.31, -18.09, 35.58, -6.8],
           [23.44, 4.5, 15.1, 9.66, 27.71],
           [11.94, 11.1 , 9.87, 9.09, 3.33]]
​
y, groups = catstack(samples)

ow.anova_oneway(y, groups)  # default is unequal variance
<class 'statsmodels.stats.base.HolderTuple'>
statistic = 1.22114568638392
pvalue = 0.3593865924697166
df = (2.0, 5.957096891952056)
df_num = 2.0
df_denom = 5.957096891952056
nobs_t = 15.0
n_groups = 3
means = array([ 0.116, 16.082,  9.066])
nobs = array([5., 5., 5.])
vars_ = array([1451.60743,   91.42492,   11.48643])
use_var = 'unequal'
welch_correction = True
tuple = (1.22114568638392, 0.3593865924697166)

# scipy
#    >>> f_oneway(*samples, equal_var=False)
#    F_onewayResult(statistic=1.22114568638392, pvalue=0.3593865924697167)

from scipy import stats
stats.f_oneway(*samples)
F_onewayResult(statistic=0.6179402863180592, pvalue=0.555354681344045)

ow.anova_oneway(y, groups, use_var="equal")
<class 'statsmodels.stats.base.HolderTuple'>
statistic = 0.6179402863180593
pvalue = 0.555354681344045
df = (2.0, 12.0)
df_num = 2.0
df_denom = 12.0
nobs_t = 15.0
n_groups = 3
means = array([ 0.116, 16.082,  9.066])
nobs = array([5., 5., 5.])
vars_ = array([1451.60743,   91.42492,   11.48643])
use_var = 'equal'
welch_correction = True
tuple = (0.6179402863180593, 0.555354681344045)
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3895,34 +3925,68 @@ def f_oneway(*samples, axis=0):
+        )
 
-    prob = special.fdtrc(dfbn, dfwn, f)   # equivalent to stats.f.sf
+        prob = 1 - stats.f.cdf(f, k - 1, df)
```
```suggestion
        prob = stats.f.sf(f, k - 1, df)
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3818,6 +3834,20 @@ def f_oneway(*samples, axis=0):
     >>> f_oneway(tillamook, newport, petersburg, magadan, tvarminne)
     F_onewayResult(statistic=7.121019471642447, pvalue=0.0002812242314534544)
 
+    Welch ANOVA will be performed if sample variances are not equal
+    and `equal_var` is False. The sample variances can be checked by
```
```suggestion
    Welch ANOVA will be performed if `equal_var` is False.
    The sample variances can be checked by
```
ISTM that we are only checking whether `equal_var` is False, not the sample variances. If the two are *equivalent* when the sample variances happen to be equal, that's fine; we still don't have to mention it.

Following up on https://github.com/scipy/scipy/pull/22152#issuecomment-2619482992, I'd prefer that we not mention variance pretesting. Re:

> The frequent recommendation for t-test is to use unequal variance Welch version by default or if there is some prior suspicion that variances are not equal, but not to rely on a rejection decision of a pre-test.
> ...
> In (very) small samples, Welch t-test and Welch anova also might not be very good because the group variance estimates are noisy and we don't pool across groups. So, assuming equal variances might still be the lesser of the two evils.

If we are not confident about a recommendation, let's not make it. We can always include statements in the examples like "Suppose that the population variances are known to be ___" and based on that, specify `equal_var = False` or leave it at the default. I don't think we need two examples; I think we can trust that our users know how to pass a boolean argument.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3735,6 +3735,14 @@ def f_oneway(*samples, axis=0):
+    equal_var: bool, optional
+        If True (default), perform a standard one-way ANOVA test that
+        assumes equal population variances [2]_.
+        If False, perform Welch's ANOVA test, which does not assume
+        equal population variances [4]_. Modified F-statistics and
+        dof will be calculated by each samples [5]_
```
This suggested that the F-statistic calculation is performed according to [5], and perhaps the p-value is determined according to [4]? I don't really understand the need for both references.

mdhaber reviewed (commented):
@YongcaiHuang I wanted to confirm that you are planning to continue with this? If so, I thought I'd coordinate with you to plan for review and revisions. It's better if there's a week or so when we will both have the time to iterate. LTM if now is good or you would like to come back later.

Some things we'll need to work on:
- [ ] Input validation for `equal_var`. Check that it's boolean, not just truthy, and if not raise. Test error message.
- [ ] Support `axis` parameter. This function accepts N-d arrays, and the statistic is calculated along `axis`. No need to add unit tests; we already have a generic property-based tests (in `tests/test_axis_nan_policy.py`) that will be easy to add. I can explain that later.
- [ ] Is the `equal_var=False` calculation really _completely_ separate from the existing one, or can some code be shared? Don't change anything now. If you avoided using the existing code for a reason, just LMK what that was.

YongcaiHuang commented:
@mdhaber thanks for your coordinate. I have read all you comments and they make sense to me. I am more avalible next week and I will implement related revision suggestions. please correct me if I am doing wrong and suggest how can we can proceed if neccessary (it's the first time for me contributing to this community)

josef-pkt commented:
FYI: I found some articles on pre-testing, two of them are at
https://github.com/statsmodels/statsmodels/issues/9496

YongcaiHuang committed:
Merge branch 'scipy:main' into main

YongcaiHuang committed:
Apply suggestions from code review

YongcaiHuang committed:
check equal_var type

YongcaiHuang committed:
support the axis parameter

YongcaiHuang committed:
Merge branch 'scipy:main' into main

YongcaiHuang commented:
@mdhaber I have finished the related implementation. feel free to review the changes

#### code review last time

1. to avoid any confusion. I removed the reference [5], only remain original paper reference [4]. 
>If False, perform Welch's ANOVA test, which does not equal population variances. Modified F-statistics and dof will be calculated by each samples [4]_.

2. thanks to the opinion on the pre-testing. I have removed related description in the documentation
3. prob calculated is changed to SF instead of 1 - CDF


#### previous plans/TODOs have been implemented

1. add instance check of boolean value for `equal_var`
2. `axis` parameter has been supported. You may find some modification regarding the array shape manipulation, but note that the calculation has not changed and result is checked as below

```python
>>> arr = np.array([[[ 141.12418768,    2.00078604,   42.08573332],
                     [ 132.21269875,   59.76185568,   -1.95455576],
                     [  62.70583556,   -6.35700275,   -5.9866934 ],
                     [  14.78154607,    1.72852285,   68.35085483],
                     [  63.16613119,   11.19410152,    0.44386323]],

                    [[   5.00511491,   80.68026995,   -2.66705743],
                     [  13.46191117,  -72.59813784, -194.027226  ],
                     [  45.09968309,    6.05105339,  -51.20938641],
                     [ 108.94822195,   -5.8174627 ,    3.52340583],
                     [  -9.92074405,  121.08955793,   23.50974032]]])

>>> F, p = stats.f_oneway(*arr, equal_var=False, axis=0)
>>> F
array([2.5054385 , 0.11489549, 2.36634122])
>>> p
array([0.1526171 , 0.74852614, 0.18344525])

>>> F, p = stats.f_oneway(*arr, equal_var=False, axis=1)
>>> F
array([0.48011662, 4.25769208, 0.21574283, 0.02982258, 0.20749929])
>>> p
array([0.53259519, 0.11990742, 0.66732487, 0.87348696, 0.68049504])
```

3. For calculation part, it indeed separated from the existing one, and  I think there is no more code can be shared. Existing one calculates the mean of Array to calculate the sum of square of the total and sum of the samples (concatenate all data into 1D via `alldata = np.concatenate(samples, axis=axis)`). However, welch one calculate the weighted means and variance for each sample. Below formula FYI


$w_k = \frac{n_k}{s^2_k}$

$\overline X^\prime = \frac{\sum w_k \overline X_k}{\sum{w_k}}$

$F^{\prime\prime} = \frac{\frac{\sum w_k (\overline X_k - \overline X ^\prime)}{k-1}} {1+ \frac{2(k-2)}{k^2 - 1}\sum(\frac{1}{n_k - 1})(1- \frac{w_k}{\sum w_k})^2}$

$df^\prime = \frac{k^2-1} {3\sum (\frac{1}{n_k-1})(1-\frac{w_k}{\sum w_k})^2}$

> @YongcaiHuang I wanted to confirm that you are planning to continue with this? If so, I thought I'd coordinate with you to plan for review and revisions. It's better if there's a week or so when we will both have the time to iterate. LTM if now is good or you would like to come back later.
> 
> Some things we'll need to work on:
> 
> * [ ]  Input validation for `equal_var`. Check that it's boolean, not just truthy, and if not raise. Test error message.
> * [ ]  Support `axis` parameter. This function accepts N-d arrays, and the statistic is calculated along `axis`. No need to add unit tests; we already have a generic property-based tests (in `tests/test_axis_nan_policy.py`) that will be easy to add. I can explain that later.
> * [ ]  Is the `equal_var=False` calculation really _completely_ separate from the existing one, or can some code be shared? Don't change anything now. If you avoided using the existing code for a reason, just LMK what that was.

YongcaiHuang committed:
retrieve doc example (deleted by accidently before)

YongcaiHuang committed:
fix f_oneway doc rendering

mdhaber commented:
Great! If you have time, consider adding `f_oneway` with keyword arguments `equal_var=False` to the test cases at `tests/test_axis_nan_policy.py` (https://github.com/scipy/scipy/blob/df134eab5a500c2146ed4552c8674a78d8154ee9/scipy/stats/tests/test_axis_nan_policy.py#L64).  To make the tests run faster locally (in case debugging is needed), consider commenting out the other test cases. If you have trouble adding it or run into errors that you have trouble fixing, LMK.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3748,6 +3748,14 @@ def f_oneway(*samples, axis=0):
+        If True (default), perform a standard one-way ANOVA test that
+        assumes equal population variances [2]_.
+        If False, perform Welch's ANOVA test, which does not assume
+        equal population variances. Modified F-statistics and
+        dof will be calculated by each samples [4]_.
```
```suggestion
        equal population variances [4]_. 
```
I think this is clear enough, right? The statistic and DOF are just part of performing Welch's test, right?

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,88 @@ def f_oneway(*samples, axis=0):
+    # the related issue: https://github.com/scipy/scipy/issues/11122
+    else:
+        # calculate basic statistics for each sample
+        samples_arrs = np.array(samples)
```
Just to disambiguate here and in future comments: I use the word "sample" (singular) to refer to a set of observations. This function accepts an arbitrary number of samples, each with a (potentially different) number of observations, in a tuple `samples`.

This won't work when different samples have different numbers of observations, right? There is no need to convert the `samples` tuple to an array, so let's leave it as a tuple.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,88 @@ def f_oneway(*samples, axis=0):
+        # calculate basic statistics for each sample
+        samples_arrs = np.array(samples)
+
+        # adjust the shape of the data based on the arrary dimension for the calculation
+        # along with input axis
+        if samples_arrs.ndim <= 2:
+            samples_arrs = samples_arrs.T
+        else:
+            samples_arrs = samples_arrs.transpose(
+                *range(1, samples_arrs.ndim - 1), 0, -1
+            )
```
I don't think you'll need this if you leave the samples in a tuple, right?

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,88 @@ def f_oneway(*samples, axis=0):
+                *range(1, samples_arrs.ndim - 1), 0, -1
+            )
+
+        means = samples_arrs.mean(axis=axis)
```
Just do something like:
```python3
means = [np.mean(sample, axis=axis) for sample in samples]
```
Same below; please just use list comprehensions instead of `apply_along_axis`.

After the reduction (of $n$ observations to a single number), you can put the results in an array if convenient.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,88 @@ def f_oneway(*samples, axis=0):
+        mean_prime = (np.sum(ws * means, axis=axis, keepdims=True) /
+                      np.sum(ws, axis=axis, keepdims=True))
+
+        numerator = (np.sum(ws * (means - mean_prime)**2, axis=axis, keepdims=True)
```
Not required, but I think it's helpful to call out equation numbers from the reference and/or Table numbers and references to particular columns. For instance, this looks like the numerator of `[4] Equation 29`?

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,88 @@ def f_oneway(*samples, axis=0):
+        df = (
+                (k ** 2 - 1) /
+                (3 * np.sum((1 / (ns - 1)) *
+                            (1 - ws / np.sum(ws, axis=axis, keepdims=True)) **2,
```
Surprisingly, our linter is not enforcing these rules right now, but:
```suggestion
                            (1 - ws / np.sum(ws, axis=axis, keepdims=True))**2,
```
PEP8 would require equal numbers of spaces before and after, and it also recommends eliminating the space for operators of high priority. Exponentiation has the highest priority, so omit the spaces, please (here and throughout).

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,88 @@ def f_oneway(*samples, axis=0):
+        prob = stats.f.sf(f, k - 1, df)
 
-    prob = special.fdtrc(dfbn, dfwn, f)   # equivalent to stats.f.sf
+        f = np.squeeze(f)
+        prob = np.squeeze(prob)
```
This is typically not required; perhaps the last reducing operations should not have `keepdims=True`? If it *is* required, specify which is the singleton `axis` to be eliminate. 

Adding the test case to `test_axis_nan_policy.py` should catch bugs like this.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7927,6 +7927,31 @@ def test_basic(self):
+        # toy samples with unequal variances
+        samples = [[-50.42, 40.31, -18.09, 35.58, -6.8],
+                   [23.44, 4.5, 15.1, 9.66, 27.71],
+                   [11.94, 11.1 , 9.87, 9.09, 3.33]]
```
Please add a test with different number of observations in each sample (or just replace this one, since I don't think there's any special treatment for equal numbers of observations in each sample).

mdhaber reviewed (commented):
@YongcaiHuang Thanks for the updates. Here are a few more comments to address. If you're able to address these and get tests in `_test_axis_nan_policy.py` to pass, I think we'll only need one more round of review after this. (During that one, I'll actually compare the equations against the reference; I haven't done that yet.)

Please ping me when this is ready for another look.

YongcaiHuang committed:
2nd review: axis parameter support; lint; test for diff. observations; test in test_axis_nan_policy.py

YongcaiHuang committed:
correct the dof calculation

YongcaiHuang commented:
hi @mdhaber, I have implemented the review comments and updated the code. let me know if there are any more comments.

1. added different number of observations in `test_stats.py`
2. improved way of supporting `axis` parameter usnig tuple and list comprehensions
3. made lint improvement (power operation `**`)
4. added test case in `test_axis_nan_policy.py` (please comment if it is in a right way)
5. added more detailed comment about calculation and equation references

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3959,52 +3959,44 @@ def f_oneway(*samples, axis=0, equal_var=True):
-        samples_arrs = np.array(samples)
+        means = np.asarray([np.mean(sample, axis=axis) for sample in samples])
+        ns = np.asarray(
+            [np.apply_along_axis(len, axis=axis, arr=sample) for sample in samples])
```
```suggestion
            [sample.shape[axis] for sample in samples])
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3959,52 +3959,44 @@ def f_oneway(*samples, axis=0, equal_var=True):
-        mean_prime = (np.sum(ws * means, axis=axis, keepdims=True) /
-                      np.sum(ws, axis=axis, keepdims=True))
+        # ref.[4] p.330 (notation is $$\hat y$$ in ref)
+        mean_prime = np.sum(ws * means, axis=0) / np.sum(ws, axis=0)
```
```suggestion
        mean_prime = np.vecdot(ws, means, axis=0) / np.sum(ws, axis=0)
```
Please test this locally before committing. 
This probably doesn't make much difference for only a few samples, but `vecdot(x, y)` can be much, much faster than `sum(x * y)` when there are many elements.

mdhaber commented:
Let's let the tests run. If they pass, this is is looking close, and I'll finish this last round of review shortly.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,80 @@ def f_oneway(*samples, axis=0):
+    else:
+        # calculate basic statistics for each sample
+        means = np.asarray([np.mean(sample, axis=axis) for sample in samples])
+        ns = np.asarray(
+            [np.apply_along_axis(len, axis=axis, arr=sample) for sample in samples])
```
```suggestion
        ns = np.asarray([sample.shape[axis] for sample in samples])
```

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7927,6 +7927,37 @@ def test_basic(self):
+                   [23.44, 4.5, 15.1, 9.66],
+                   [11.94, 11.1 , 9.87, 9.09, 3.33]]
+
+        with assert_raises(TypeError):
+            stats.f_oneway(*samples, equal_var="False")
```
Please make this a separate test - this part is input validation of the `equal_var` parameter; the rest of the test is accuracy of the Welch test. We also try to check for the particular error message these days (`match` parameter).

```python3
def test_equal_var_input_validation(self):
    message = "Expected a boolean value for 'equal_var'"
    with pytest.raises(TypeError, match=message):
        stats.f_oneway(*samples, equal_var="False")
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,80 @@ def f_oneway(*samples, axis=0):
-    msw = sswn / dfwn
-    with np.errstate(divide='ignore', invalid='ignore'):
-        f = msb / msw
+    # if population variances are not equeal, perform Welch ANOVA
+    # the related issue: https://github.com/scipy/scipy/issues/11122
```
First sentence doesn't describe what the code does; the condition is really just whether `equal_var=True`.
Also, we don't typically link to enhancement requests in the code (although it's not a bad idea to link to bug reports in tests). If someone is interested, they can look at the blame for these lines, and from there it's easy to see the PR, and from there the original issue.
```suggestion
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,80 @@ def f_oneway(*samples, axis=0):
+
+        vars_ = np.asarray([np.var(sample, axis=axis, ddof=1) for sample in samples])
+
+        # calculate weight by number of data and varianc
```
```suggestion
        # calculate weight by number of data and variance
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3809,6 +3817,10 @@ def f_oneway(*samples, axis=0):
            http://www.biostathandbook.com/onewayanova.html
 
+    .. [4] B. L. Welch, "On the Comparison of Several Mean Values:
+           An Alternative Approach,"Biometrika, vol. 38, no. 3/4,
```
```suggestion
           An Alternative Approach", Biometrika, vol. 38, no. 3/4,
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,80 @@ def f_oneway(*samples, axis=0):
+        numerator = np.sum(ws * (means - mean_prime)**2, axis=0) / (k - 1)
+        denominator = (
+                1 + 2 * (k - 2) / (k**2 - 1) *
+                np.sum((1 / (ns - 1)) *
```
Please create a variable e.g. `f_t = ns - 1` and cite where in the paper the definition comes from.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,80 @@ def f_oneway(*samples, axis=0):
+        denominator = (
+                1 + 2 * (k - 2) / (k**2 - 1) *
+                np.sum((1 / (ns - 1)) *
+                       (1 - ws / np.sum(ws, axis=0))**2,
```
```suggestion
                       (1 - ws / S_ws)**2,
```
Let's create a variable `S_ws = np.sum(ws, axis=0)` and use it here and below. I think it will make the equations ever so slightly easier to compare against the reference, and it's a little faster.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,80 @@ def f_oneway(*samples, axis=0):
+    # the related issue: https://github.com/scipy/scipy/issues/11122
+    else:
+        # calculate basic statistics for each sample
+        means = np.asarray([np.mean(sample, axis=axis) for sample in samples])
```
If you're willing, it would be super helpful to 1) replace the names with the symbols used in the paper and 2) identify where in the paper these symbols are defined. For instance, it took me about a minute to:
- identify where `means` is used in the equations below so that I can see that it's $y_t$ in the paper, and then
- find where it is defined at the beginning of the second paragraph. 

If I have to do that for every symbol, it adds up.

```suggestion
        # Beginning of second paragraph [4] page 1:
        # "As a particular case $y_t$ may be the means ... of samples
        y_t = np.asarray([np.mean(sample, axis=axis) for sample in samples])
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,80 @@ def f_oneway(*samples, axis=0):
+        means = np.asarray([np.mean(sample, axis=axis) for sample in samples])
+        ns = np.asarray(
+            [np.apply_along_axis(len, axis=axis, arr=sample) for sample in samples])
+        k = len(samples)
```
```suggestion
        # "... from $k$ different normal populations..."
        k = len(samples)
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,80 @@ def f_oneway(*samples, axis=0):
+        ns = np.asarray(
+            [np.apply_along_axis(len, axis=axis, arr=sample) for sample in samples])
+        k = len(samples)
+
+        vars_ = np.asarray([np.var(sample, axis=axis, ddof=1) for sample in samples])
```
```suggestion
        # "The separate samples provide estimates $s_t^2$ of the $\sigma_t^2$."
        s_t2= np.asarray([np.var(sample, axis=axis, ddof=1) for sample in samples])
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,80 @@ def f_oneway(*samples, axis=0):
+        vars_ = np.asarray([np.var(sample, axis=axis, ddof=1) for sample in samples])
+
+        # calculate weight by number of data and varianc
+        # ref.[4] p.330 ($$w = 1/{\lambda s^2}$$ where $$\lambda=1/n$$ in ref)
+        ws = ns / vars_
```
```suggestion
        # "we have $\lambda_t = 1 / n_t$ ... where w_t = 1 / {\lambda_t s_t^2}$"
        w_t = n_t / s_t2
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,80 @@ def f_oneway(*samples, axis=0):
+        ws = ns / vars_
+
+        # calculate adjusted grand mean
+        # ref.[4] p.330 (notation is $$\hat y$$ in ref)
+        mean_prime = np.sum(ws * means, axis=0) / np.sum(ws, axis=0)
```
```suggestion
        # "... and $\hat{y} = \sum w_t y_t / \sum w_t$..."
        y_hat = np.sum(w_t * y_t, axis=0) / np.sum(w_t, axis=0)
```

mdhaber reviewed (commented):
This is great! I checked the math against the paper, and it all looks good. I ran the R code, and the test compares against that output faithfully. The `test_axis_nan_policy` test looks right, and that makes sure that all the `stats` bells and whistles (`axis`, `nan_policy`, `keepdims`, etc.) are working.

Please address https://github.com/scipy/scipy/pull/22152/files#r1941889786 and the new comments, and I think this will be good to go!

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -3908,34 +3922,80 @@ def f_oneway(*samples, axis=0):
+        # adjusted degree of freedom
+        # ref.[4] p.334 eq.30
+        df = (
+                (k**2 - 1) /
+                (3 * np.sum((1 / (ns - 1)) *
+                            (1 - ws / np.sum(ws, axis=0))**2, axis=0))
```
Also, technically PEP8 suggest that operators should be on the next line:
https://peps.python.org/pep-0008/#should-a-line-break-before-or-after-a-binary-operator

But it's OK with me for you to leave it as-is. You've clearly been very careful to make this readable, so it's a lot better than most mathy code.

mdhaber reviewed (approved):
Thanks again @YongcaiHuang. I'm indicating my approval here; please just address the remaining comments and I can merge. With the exception of https://github.com/scipy/scipy/pull/22152/files#r1947499618, these are mostly style suggestions, so if you feel strongly against any, just let me know why, and I think I can merge without them being changed. (After all, if I really feel strongly enough about any of these, I could change it myself later.)

YongcaiHuang committed:
Update scipy/stats/_stats_py.py

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>

YongcaiHuang committed:
Apply suggestions from code review

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>

YongcaiHuang committed:
review update

YongcaiHuang reviewed (commented):
@mdhaber thanks for the suggestion. I have renamed related variables to be consistent to the paper notation. Also, boolean validation test has been separated.

However, I'd like to raise problem when using
```python
n_t = np.asarray([sample.shape[axis] for sample in samples])
```

there is no problem if input samples are in 1D, but when passing input more than 1D the shape is different with `s_t2` so that `w_t`, `f`, `hat_f2` cannot be correctly calculated. find the e.g. below

```python
# input samples are in 2D
arr = np.array([[[ 141.12418768,    2.00078604,   42.08573332],
                 [ 132.21269875,   59.76185568,   -1.95455576],
                 [  62.70583556,   -6.35700275,   -5.9866934 ],
                 [  14.78154607,    1.72852285,   68.35085483],
                 [  63.16613119,   11.19410152,    0.44386323]],

                [[   5.00511491,   80.68026995,   -2.66705743],
                 [  13.46191117,  -72.59813784, -194.027226  ],
                 [  45.09968309,    6.05105339,  -51.20938641],
                 [ 108.94822195,   -5.8174627 ,    3.52340583],
                 [  -9.92074405,  121.08955793,   23.50974032]]])

F, p = stats.f_oneway(*arr, equal_var=False, axis=0)

...

n_t = np.asarray([sample.shape[axis] for sample in samples])
s_t2 = np.asarray([np.var(sample, axis=axis, ddof=1) for sample in samples])
# n_t = array([5, 5])
# s_t2 = array([[2814.77499297,  702.60985223, 1090.90255875],
#               [2230.2544843 , 5790.94734999, 7771.11087815]])

w_t = n_t / s_t2
# ValueError: operands could not be broadcast together with shapes (2,) (2,3)
```

instead, I am trying to keep `n_t` with the same shape as other basic statistics
```python
n_t = np.asarray(
              [np.apply_along_axis(len, axis=axis, arr=sample) for sample in samples])
# n_t = array([[5, 5, 5],
#              [5, 5, 5]])
# following statistics can be calculated
```

I kept second way to pass the test. Do you have any better recommendations? feel free to suggest.

mdhaber commented:
Ah, I see. Well, I think one option is to reshape the result: 
```python3
n_t = np.reshape(n_t, y_t.shape)
```
I think we would use this approach if we were to rewrite this in terms of the [Array API Standard](https://data-apis.org/array-api/latest/) because `apply_along_axis` is not in the standard. I'll consider whether to change it during my final review. Hopefully I can merge today!

Did you plan to continue with Games Howell?

YongcaiHuang committed:
reshape n_t (number of observation)

YongcaiHuang commented:
Thanks for the suggestion! No problem for 1D and ND arrary on local testing. code has been updated. It looks all good now. Again, thanks for your coordination.👍 

For Games Howell, I'd love to continue, but I got no time recently. If no rush, I can update whenever I got time. How do you think?

mdhaber commented:
That's fine. We discussed that the next step there was to show me what it would look like if you were to *change* the existing code to only do Games-Howell, just as a draft so I can see the difference. Feel free to do that when you have time, and then we can talk about how to add an `equal_var` parameter so the user can select one or the other.
